File: tmdb-core/src/main/java/dev/reuss/tmdb/domain/tv/episode/model/TvEpisodeCredits.java
package dev.reuss.tmdb.domain.tv.episode.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import dev.reuss.tmdb.common.TmdbModel;
import dev.reuss.tmdb.common.credit.CastCredit;
import dev.reuss.tmdb.common.credit.CreditsResponse;
import dev.reuss.tmdb.common.credit.CrewCredit;
import dev.reuss.tmdb.common.credit.GuestStarCredit;
import dev.reuss.tmdb.util.TmdbCollections;

import java.util.List;

/**
 * Credits for a TMDB TV episode.
 *
 * @param cast       episode cast credits
 * @param crew       episode crew credits
 * @param guestStars episode guest stars
 * @param id         TMDB episode id
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record TvEpisodeCredits(
        int id,
        List<CrewCredit> crew,
        List<CastCredit> cast,

        @JsonProperty("guest_stars")
        List<GuestStarCredit> guestStars

) implements CreditsResponse<CastCredit, CrewCredit>, TmdbModel {

    public TvEpisodeCredits {
        cast = TmdbCollections.list(cast);
        crew = TmdbCollections.list(crew);
        guestStars = TmdbCollections.list(guestStars);
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/domain/tv/episode/model/TvEpisodeCredits.kt
package dev.reuss.tmdb.domain.tv.episode.model

import com.fasterxml.jackson.annotation.JsonProperty
import dev.reuss.tmdb.common.TmdbModel
import dev.reuss.tmdb.common.credit.CastCredit
import dev.reuss.tmdb.common.credit.CreditsResponse
import dev.reuss.tmdb.common.credit.CrewCredit
import dev.reuss.tmdb.common.credit.GuestStarCredit

/**
 * Credits for a TMDB TV episode.
 *
 * @property cast episode cast credits
 * @property crew episode crew credits
 * @property guestStars episode guest stars
 * @property id TMDB episode id
 */
@JvmRecord
data class TvEpisodeCredits(
    @all:JsonProperty("id")
    val id: Int,
    @all:JsonProperty("crew")
    override val crew: List<CrewCredit> = emptyList(),
    @all:JsonProperty("cast")
    override val cast: List<CastCredit> = emptyList(),
    @all:JsonProperty("guest_stars")
    val guestStars: List<GuestStarCredit> = emptyList(),
) : CreditsResponse<CastCredit, CrewCredit>,
    TmdbModel
